llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/IREEPyDM/Transforms/RTL/LinkageAnalysis.h

```diff
@@ -0,0 +1,39 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_LLVM_EXTERNAL_PROJECTS_IREE_DIALECTS_DIALECT_IREEPYDM_TRANSFORMS_RTL_LINKAGE_ANALYSIS_H
+#define IREE_LLVM_EXTERNAL_PROJECTS_IREE_DIALECTS_DIALECT_IREEPYDM_TRANSFORMS_RTL_LINKAGE_ANALYSIS_H
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/DenseSet.h"
+
+namespace mlir {
+class Operation;
+
+namespace iree_pydm {
+
+/// An analysis of the external linkage which must be satisfied.
+class LinkageAnalysis {
+ public:
+  LinkageAnalysis(Operation *moduleOp);
+
+  /// Whether there are any external functions that need resolution.
+  bool hasExternFuncs() const { return !externFuncOps.empty(); }
+
+  /// Gets a next extern func from the set.
+  Operation *getNextExternFunc() const { return *externFuncOps.begin(); }
+
+  /// Erases an operation.
+  void erase(Operation *externOp) { externFuncOps.erase(externOp); }
```

Comment:
Usually analysis isn't mutable like this and it took me a bit of time to walk through the code and understand how this was happening (the 1000 iteration loop with the check for any and then the erase deep down during inlining). Maybe rename to be more explicit as to what it is - like `UnresolvedExternSet`? It's simple enough that it could just live at the top of LinkRTLPass.cpp (since it's not a general analysis that would be reusable in any other pass it keeps it more contained).

Reply:
(llvm analysis being mutable is more of a bug than a feature in most cases - immutability is what allows for parallelism in MLIR - here this instance is only used in a module pass but erring on the side of immutability is good in other ones :)